www/proxy-sso/src/opnsense/mvc/app/controllers/OPNsense/ProxySSO/Api/ServiceController.php

```diff
@@ -0,0 +1,261 @@
+            }
+
+            $host_esc = escapeshellarg("{$ldap_server->host}");
+            $output = array("# ping -c 1 -W 1 {$host_esc}");
+            $retval = 0;
+            exec("ping -c 1 -W 1 {$host_esc}", $output, $retval);
+            $ldap_server_ping = [ "status" => $retval == 0 ? "ok" : "failure"];
+            $ldap_server_ping["dump"] = implode("\n", $output);
+        }
+
+        // DNS
+        $dns_server = array();
+        $nameservers = preg_grep('/nameserver/', file('/etc/resolv.conf'));
+        $dns_servers = array();
+        foreach($nameservers as $key => $record) {
+            $parts = explode(' ', $record);
+            $dns_servers[] = trim($parts[1]);
+        }
+        $dns_server = [ "status" => count($dns_servers) ? "ok" : "failure"];
+        if(!count($dns_servers)) {
+            $dns_server["message"] = gettext("DNS server not found");
+        }
+        $output = array("# cat /etc/resolv.conf");
+        exec('cat /etc/resolv.conf', $output);
+        $dns_server["dump"] = implode("\n", $output);
+
+        // DNS: hostname
+        $resolv_direct = chop(shell_exec("drill {$hostname} | grep -A 1 'ANSWER SECTION' | tail -n 1 | awk '{print \$5}'"));
+        $dns_hostname_resolution = [ "status" => !empty($resolv_direct) && filter_var($resolv_direct, FILTER_VALIDATE_IP) ? "ok" : "failure"];
+        $output = array("# drill {$hostname}");
+        exec("drill {$hostname}", $output);
+        $dns_hostname_resolution["dump"] = implode("\n", $output);
+        
+        $resolv_reverse = null;
+        $dns_hostname_reverse_resolution = array();
+        if(!empty($resolv_direct) && filter_var($resolv_direct, FILTER_VALIDATE_IP)) {
+            $resolv_reverse = chop(shell_exec("drill -x {$resolv_direct} | grep -A 1 'ANSWER SECTION' | tail -n 1 | awk '{print \$5}'"));
+            if(strtolower($resolv_reverse) != strtolower("{$hostname}.")) {
+                $dns_hostname_reverse_resolution["message"] = gettext("Hostname doesn't resolved to host IP.");
+            }
+        }
+        else {
+            $dns_hostname_reverse_resolution["message"] = gettext("Hostname doesn't resolved to IP.");
+        }
+        $dns_hostname_reverse_resolution["status"] = strtolower($resolv_reverse) == strtolower("{$hostname}.") ? "ok" : "failure";
+        $output = array("# drill -x {$resolv_direct}");
+        exec("drill -x {$resolv_direct}", $output);
+        $dns_hostname_reverse_resolution["dump"] = implode("\n", $output);
+
+
+        // DNS: LDAP server
+        ldap_dns:
+        $dns_ldap_reverse_resolution = array( "status" => "failure" );
+        if(empty($ldap_ip)) {
+            $dns_ldap_reverse_resolution["message"] = gettext("Unknown LDAP server IP.");
+        }
+        else {
+            $ldap_ip_esc = escapeshellarg($ldap_ip);
+            $resolv_reverse = chop(shell_exec("drill -x {$ldap_ip_esc} | grep -A 1 'ANSWER SECTION' | tail -n 1 | awk '{print \$5}'"));
```

Comment:
if you need this more often, you may want to extract that to a function and do all the escaping there which may also improve security (less likely to forget to escape something).